### PR TITLE
TOKENS: sPrime, vPrime, PRIME

### PR DESCRIPTION
**Pull request by kavsky02**

### CODE REVIEW

File: contracts/Pool.sol [hunk 1]
```diff
@@ -40,10 +44,64 @@ contract Pool is OwnableUpgradeable, ReentrancyGuardUpgradeable, IERC20 {
...
+        for (uint i = 0; i < locks[account].length; i++) {
+            if (locks[account][i].unlockTime > block.timestamp) {
+                lockedBalance += locks[account][i].amount;
+            }
+        }
+        return lockedBalance;
+    }
+
+    function getNotLockedBalance(address account) public view returns (uint256) {
+        return balanceOf(account) - getLockedBalance(account);
```
Comment (mbare0) [inline contracts/Pool.sol:72]: is there a possible scenario when getLockedBalance(account) > balanceOf(account)? Maybe it's good to add a simple check for that and return 0

File: contracts/Pool.sol [hunk 2]
```diff
@@ -40,10 +44,64 @@ contract Pool is OwnableUpgradeable, ReentrancyGuardUpgradeable, IERC20 {
...
+    }
+
+
+    function lockDeposit(uint256 amount, uint256 lockTime) public {
+        uint256 lockedBalance = getLockedBalance(msg.sender);
+        require(balanceOf(msg.sender) - lockedBalance >= amount, "Insufficient balance to lock");
```
Comment (mbare0) [inline contracts/Pool.sol:78]: we can use `getNotLockedBalance` here

File: contracts/Pool.sol [hunk 3]
```diff
@@ -283,11 +351,18 @@ contract Pool is OwnableUpgradeable, ReentrancyGuardUpgradeable, IERC20 {
         tokenAddress.safeTransfer(to, amount);
     }
 
+    function isWithdrawalAmountAvailable(address account, uint256 amount) public view returns (bool) {
+        uint256 lockedBalance = getLockedBalance(account);
+        return amount <= balanceOf(account) - lockedBalance;
```
Comment (mbare0) [inline contracts/Pool.sol:356]: there is a method `getNotLockedBalance`

File: contracts/Pool.sol [hunk 4]
```diff
@@ -40,10 +44,64 @@ contract Pool is OwnableUpgradeable, ReentrancyGuardUpgradeable, IERC20 {
...
+    function getFullyVestedLockedBalance(address account) public view returns (uint256 fullyVestedBalance) {
+        fullyVestedBalance = 0;
+        for (uint i = 0; i < locks[account].length; i++) {
+            if (locks[account][i].unlockTime > block.timestamp) { // Lock is still active
+                fullyVestedBalance += locks[account][i].amount * locks[account][i].lockTime / MAX_LOCK_TIME;
+            }
+        }
+    }
+
+    function setVPrimeController(vPrimeController _vPrimeController) public onlyOwner {
```
Comment (mbare0) [inline contracts/Pool.sol:101]: shouldn't we take it from TokenManager?

Comment (mbare0) [reply contracts/Pool.sol:101]: to have a single point of truth

File: contracts/facets/arbitrum/ConvexFacetArbi.sol [hunk 5]
```diff
@@ -4,6 +4,7 @@ pragma solidity 0.8.17;
 
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 
+import "@uniswap/lib/contracts/libraries/TransferHelper.sol";
```
Comment (mbare0) [inline contracts/facets/arbitrum/ConvexFacetArbi.sol:7]: why is it needed? (and in files below)

File: contracts/interfaces/IPool.sol [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    function name() external view returns (string memory);
+    function totalSupplyCap() external view returns (uint256);
+    function ratesCalculator() external view returns (address);
+    function borrowersRegistry() external view returns (address);
+    function poolRewarder() external view returns (address);
+    function depositIndex() external view returns (address);
+    function borrowIndex() external view returns (address);
+    function tokenAddress() external view returns (address);
+    function vestingDistributor() external view returns (address);
+    function vPrimeControllerContract() external view returns (address);
```
Comment (mbare0) [inline contracts/interfaces/IPool.sol:37]: please add the new events as well, generally all events should be defined in interfaces for the UI/bots

File: contracts/Pool.sol [hunk 7]
```diff
@@ -40,10 +44,64 @@ contract Pool is OwnableUpgradeable, ReentrancyGuardUpgradeable, IERC20 {
...
+
+        proxyCalldata(
+            address(vPrimeControllerContract),
+            abi.encodeWithSignature("updateVPrimeSnapshot(address)", msg.sender),
+            false
+        );
+    }
+
+
+    function getFullyVestedLockedBalance(address account) public view returns (uint256 fullyVestedBalance) {
```
Comment (mbare0) [inline contracts/Pool.sol:92]: Could you please add a thorough description of this method for auditors?

File: contracts/interfaces/ISPrime.sol [hunk 8]
```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity ^0.8.17;
+
+interface ISPrime {
```
Comment (mbare0) [inline contracts/interfaces/ISPrime.sol:5]: ISPrimeTraderJoeV2

File: contracts/lib/local/DeploymentConstants.sol [hunk 9]
```diff
@@ -20,13 +20,13 @@ library DeploymentConstants {
 
     address private constant _NATIVE_ADDRESS = 0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7;
 
-    address private constant _DIAMOND_BEACON_ADDRESS = 0x8bCe54ff8aB45CB075b044AE117b8fD91F9351aB;
+    address private constant _DIAMOND_BEACON_ADDRESS = 0x821f3361D454cc98b7555221A06Be563a7E2E0A6;
```
Comment (mbare0) [inline contracts/lib/local/DeploymentConstants.sol:23]: please revert

File: contracts/tokens/PrimeBridge.sol [hunk 10]
```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: BUSL-1.1
+// Last deployed from commit: ;
+pragma solidity 0.8.17;
+
+import "@layerzerolabs/solidity-examples/contracts/token/oft/v2/BaseOFTV2.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract PrimeBridge is BaseOFTV2 {
```
Comment (mbare0) [inline contracts/tokens/PrimeBridge.sol:8]: what is this contract based on? (besides extending BaseOFTV2)

File: contracts/tokens/Prime_L2.sol [hunk 11]
```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: BUSL-1.1
+// Last deployed from commit: ;
+
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@layerzerolabs/solidity-examples/contracts/token/oft/v2/BaseOFTV2.sol";
+
+contract Prime_L2 is BaseOFTV2, ERC20 {
```
Comment (mbare0) [inline contracts/tokens/Prime_L2.sol:9]: what is this contract based on? (besides extending BaseOFTV2)

File: contracts/tokens/mock/sPrimeMock.sol [hunk 12]
```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@redstone-finance/evm-connector/contracts/core/ProxyConnector.sol";
+import {SolvencyFacetProd} from "../../facets/SolvencyFacetProd.sol";
+import "../vPrimeController.sol";
+
+contract SPrimeMock is ERC20, Ownable, ProxyConnector {
```
Comment (mbare0) [inline contracts/tokens/mock/sPrimeMock.sol:10]: shouldn't be removed?

File: contracts/tokens/mock/vPrimeMock.sol [hunk 13]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "../vPrime.sol";
+
+
+contract vPrimeMock is vPrime {
```
Comment (mbare0) [inline contracts/tokens/mock/vPrimeMock.sol:7]: shouldn't be removed?

File: contracts/tokens/sPrime.sol [hunk 14]
```diff
@@ -0,0 +1,643 @@
...
+
+// SPrime contract declaration
+contract SPrime is ISPrime, ReentrancyGuardUpgradeable, OwnableUpgradeable, ERC20Upgradeable, SolvencyMethods {
+    using SafeERC20 for IERC20Metadata; // Using SafeERC20 for IERC20 for safe token transfers
+    using LiquidityAmounts for address; // Using LiquidityAmounts for address for getting amounts of liquidity
+    using SafeCast for uint256; // Using SafeCast for uint256 for safe type casting
+    using PackedUint128Math for bytes32;
+
+    // Constants declaration
+    uint256 private constant _MAX_SIPPIAGE = 5;
```
Comment (mbare0) [inline contracts/tokens/sPrime.sol:32]: _MAX_SLIPPAGE

File: contracts/tokens/sPrime.sol [hunk 15]
```diff
@@ -0,0 +1,643 @@
...
+            amount: amount,
+            unlockTime: block.timestamp + lockPeriod
+        }));
+    }
+
+    /**
+    * @dev Releases a locked balance at a specified index.
+    * @param index The index of the lock to be released.
+    */
+    function releaseBalance(uint256 index) public {
```
Comment (mbare0) [inline contracts/tokens/sPrime.sol:613]: @kamilovsky01 vPRIME contracts miss this functionality afaik

**Comment by kavsky02** [comment]
THIS PR is outdated as of now - we're still working within the individual sPRIME and vPRIME PRs

**Pull request closed without merging**